File: Question_section_Semantic_Search/Python_Ver_Question_semantics/src/semantic_search/backend.py
import time
from datetime import datetime
from typing import Dict

from .config import (
    MONGODB_URI, DB_NAME, COLL_1, COLL_2,
    VECTOR_INDEX_1, VECTOR_INDEX_2, SIM_THRESHOLD,
    LAST_7_DAYS, DATA_PATH
)
from .utils import load_and_flatten
from .embeddings import Embedder
from .mongo_store import MongoVectorStore

class SemanticBackend:
    def __init__(self):
        self.embedder = Embedder()
        self.mongo1 = MongoVectorStore(MONGODB_URI, DB_NAME, COLL_1)
        self.mongo2 = MongoVectorStore(MONGODB_URI, DB_NAME, COLL_2)

    def build_from_dataset(self):
        docs = load_and_flatten(DATA_PATH)
        embeddings = self.embedder.encode_texts([d["question"] for d in docs])

        split = len(docs) // 2
        docs1, emb1 = docs[:split], embeddings[:split]
        docs2, emb2 = docs[split:], embeddings[split:]

        self.mongo1.upsert_many(docs1, emb1)
        self.mongo2.upsert_many(docs2, emb2)

        self.mongo1.create_ttl_index(LAST_7_DAYS)
        self.mongo2.create_ttl_index(LAST_7_DAYS)

        print(f"Inserted {len(docs1)} docs into {COLL_1}")
        print(f"Inserted {len(docs2)} docs into {COLL_2}")
        print("TTL indexes created for both collections.")

    def query_flow(self, question: str):
        query_vec = self.embedder.encode_one(question)

        print("\n" + "=" * 70)
        print("QUESTION:", question)
        print("=" * 70)

        print("\n[1] Searching collection 1...")
        res1 = self.mongo1.vector_search(query_vec, VECTOR_INDEX_1, limit=1)
        if res1 and res1[0]["score"] >= SIM_THRESHOLD:
            self.flash(res1[0], "collection 1")
            return res1[0]

        print("\n[2] Searching collection 2...")
        res2 = self.mongo2.vector_search(query_vec, VECTOR_INDEX_2, limit=1)
        if res2 and res2[0]["score"] >= SIM_THRESHOLD:
            self.flash(res2[0], "collection 2")
            return res2[0]

        print("\n[3] No match found. Saving question to collection 2...")
        new_doc = {
            "id": f"new_{int(time.time() * 1000)}",
            "question": question,
            "answer": None,
            "section_number": None,
            "section_title": "new_questions",
            "embedding": query_vec,
            "created_at": datetime.utcnow()
        }
        self.mongo2.coll.insert_one(new_doc)
        print("Saved to collection 2.")
        return None

    def flash(self, result: Dict, source: str):
        print("\n" + "🎉" * 20)
        print(f"FOUND IN {source.upper()}")
        print(f"Question: {result['question']}")
        print(f"Answer: {result['answer']}")
        print(f"Similarity: {result['score']:.3f}")
        print("🎉" * 20)

File: Question_section_Semantic_Search/NodeJs_Ver_Question_semantics/src/backend.js
import { CONFIG } from "./config.js";
import { loadAndFlatten } from "./utils.js";
import { Embedder } from "./embedder.js";
import { MongoVectorStore } from "./mongoStore.js";

export class SemanticBackend {
  constructor() {
    this.embedder = new Embedder();
    this.mongo1 = new MongoVectorStore(CONFIG.MONGODB_URI, CONFIG.DB_NAME, CONFIG.COLL_1);
    this.mongo2 = new MongoVectorStore(CONFIG.MONGODB_URI, CONFIG.DB_NAME, CONFIG.COLL_2);
  }

  async init() {
    await this.mongo1.init();
    await this.mongo2.init();
  }

  async buildFromDataset() {
    const docs = loadAndFlatten(CONFIG.DATA_PATH);
    const embeddings = await this.embedder.encodeMany(docs.map(d => d.question));

    const split = Math.floor(docs.length / 2);
    const docs1 = docs.slice(0, split);
    const emb1 = embeddings.slice(0, split);
    const docs2 = docs.slice(split);
    const emb2 = embeddings.slice(split);

    await this.mongo1.upsertMany(docs1, emb1);
    await this.mongo2.upsertMany(docs2, emb2);

    await this.mongo1.createTTLIndex(CONFIG.LAST_7_DAYS);
    await this.mongo2.createTTLIndex(CONFIG.LAST_7_DAYS);

    console.log(`Inserted ${docs1.length} docs into ${CONFIG.COLL_1}`);
    console.log(`Inserted ${docs2.length} docs into ${CONFIG.COLL_2}`);
    console.log("TTL indexes created for both collections.");
  }

  async queryFlow(question) {
    const queryVec = await this.embedder.encodeOne(question);

    console.log("\n" + "=".repeat(70));
    console.log("QUESTION:", question);
    console.log("=".repeat(70));

    console.log("\n[1] Searching collection 1...");
    const res1 = await this.mongo1.vectorSearch(queryVec, CONFIG.VECTOR_INDEX_1, 1);
    if (res1.length && res1[0].score >= CONFIG.SIM_THRESHOLD) {
      this.flash(res1[0], "collection 1");
      return res1[0];
    }

    console.log("\n[2] Searching collection 2...");
    const res2 = await this.mongo2.vectorSearch(queryVec, CONFIG.VECTOR_INDEX_2, 1);
    if (res2.length && res2[0].score >= CONFIG.SIM_THRESHOLD) {
      this.flash(res2[0], "collection 2");
      return res2[0];
    }

    console.log("\n[3] No match found. Saving question to collection 2...");
    await this.mongo2.collection.insertOne({
      id: `new_${Date.now()}`,
      question,
      answer: null,
      section_number: null,
      section_title: "new_questions",
      embedding: queryVec,
      created_at: new Date()
    });

    console.log("Saved to collection 2.");
    return null;
  }

  flash(result, source) {
    console.log("\n" + "🎉".repeat(20));
    console.log(`FOUND IN ${source.toUpperCase()}`);
    console.log(`Question: ${result.question}`);
    console.log(`Answer: ${result.answer}`);
    console.log(`Similarity: ${result.score.toFixed(3)}`);
    console.log("🎉".repeat(20));
  }
}